bqskit/runtime/detached.py

```diff
@@ -0,0 +1,571 @@
+
+    def _handle_cancel_comp_task(self, request: uuid.UUID) -> None:
+        """Cancel a compilation task in the system."""
+        self.logger.info(f'Cancelling: {request}.')
+
+        # Remove task from server data
+        mailbox_id, client_conn = self.tasks[request]  # .pop(request)
+        # self.mailbox_to_task_dict.pop(mailbox_id)
+        self.mailboxes.pop(mailbox_id)
+        self.clients[client_conn].remove(request)
+
+        # Forward internal cancel messages
+        addr = RuntimeAddress(-1, mailbox_id, 0)
+        self._handle_cancel(addr)
+
+        # Acknowledge the cancel request
+        if not client_conn.closed:
+            client_conn.send((RuntimeMessage.CANCEL, None))
+
+    def _handle_cancel(self, addr: RuntimeAddress) -> None:
+        """Cancel a runtime task in the system."""
+        for mconn in self.managers:
+            self.outgoing.append((mconn, RuntimeMessage.CANCEL, addr))
+
+    def _handle_error(self, error_payload: tuple[int, str]) -> None:
+        """Forward an error to the appropriate client and disconnect it."""
+        if not isinstance(error_payload, tuple):
+            self._handle_shutdown()
+            raise RuntimeError(error_payload)
+
+        tid = error_payload[0]
+        conn = self.tasks[self.mailbox_to_task_dict[tid]][1]
+        self.outgoing.append((conn, RuntimeMessage.ERROR, error_payload[1]))
+        self._handle_disconnect(conn)
+
+    def _handle_log(self, log_payload: tuple[int, LogRecord]) -> None:
+        """Forward logs to appropriate client."""
+        tid = log_payload[0]
+        conn = self.tasks[self.mailbox_to_task_dict[tid]][1]
+        self.outgoing.append((conn, RuntimeMessage.LOG, log_payload[1]))
+
+    def _get_new_mailbox_id(self) -> int:
+        """Unique mailbox id counter."""
+        new_id = self.mailbox_counter
+        self.mailbox_counter += 1
+        return new_id
+
+
+def parse_ipports(ipports_str: list[str]) -> list[tuple[str, int]]:
+    """Parse command line ip and port inputs."""
+    ipports = []
+    for ipport_group in ipports_str:
+        for ipport in ipport_group.split(','):
+            if ipport.strip() == '':
+                continue
+            comps = ipport.strip().split(':')
+
+            if len(comps) == 1:
+                ip, port = comps[0], '7473'
+
+            elif len(comps) == 2:
+                ip, port = comps
+
+            else:
+                raise ValueError(f'Invalid manager address: {ipport}.')
+
+            if not (0 <= int(port) < 65536):
+                raise ValueError(f'Invalid port number: {ipport}')
+
+            ipports.append((ip, int(port)))
+    return ipports
+
+
+def start_server() -> None:
+    """Entry point for a detached runtime server process."""
+    parser = argparse.ArgumentParser(
+        prog='BQSKit Server',
+        description='Launch a BQSKit runtime server process.',
+    )
+    parser.add_argument(
```

Comment:
The positional arguments are required by default. Running `bqskit-server` without any additional parameters or flags will crash with the following message:

```
usage: BQSKit Server [-h] [-p PORT] [--verbose] managers [managers ...]
BQSKit Server: error: the following arguments are required: managers
```

Also the port isn't required since it defaults to the magic :) one.